lib/charms/lego_base_k8s/v1/lego_client.py

```diff
@@ -0,0 +1,431 @@
+        being the environment variables in lowercase kebab-case, and the values being the values
+        that are required for Lego.
+
+        For example, the required variables for Cloudflare is documented in https://go-acme.github.io/lego/dns/cloudflare/index.html
+        The format for the secret of a cloudflare plugin should be
+        ```
+        {
+            "cloudflare-email": "your@email.com",
+            "cloudflare-api-key": "yourapikey123"
+        }
+        ```
+        A user can create a secret in this format by running
+        ```
+        juju add-secret plugin-credentials cloudflare-email="your@email.com" cloudflare-api-key="yourapikey123"
+        ```
+
+provides:
+  certificates:
+    interface: tls-certificates
+  send-ca-cert:
+    interface: certificate_transfer
+requires:
+  logging:
+    interface: loki_push_api
+
+parts:
+  charm:
+    build-snaps:
+      - go
+```
+
+Then, to use the library in an example charm, you can do something like the following:
+```python
+from charms.lego_client_operator.v1.lego_client import AcmeClient
+from ops.main import main
+class ExampleAcmeCharm(AcmeClient):
+    def __init__(self, *args):
+        super().__init__(*args, plugin="cloudflare")
+        self._server = "https://acme-staging-v02.api.letsencrypt.org/directory"
+        self._email = "testingmctestface@test.com"
+
+    def _validate_plugin_config_options(self, plugin_config: Dict[str, str]) -> str:
+        if "cloudflare-email" not in plugin_config:
+            return "API user was not provided"
+        if "cloudflare-api-key" not in plugin_config:
+            return "API key was not provided"
+        return ""
+```
+
+The minimum that Charms such as the above example are expected to do is:
+- Inherit from AcmeClient,
+- Call `super().__init__(*args, plugin="<plugin-name>")` with the lego plugin name,
+  you can look at the list of plugins at https://go-acme.github.io/lego/dns/index.html#dns-providers
+- Implement the `_validate_plugin_config_options` method,
+  The function should match the signature of the example charm above, and should
+  validate the plugin specific configuration. It should return a string with an
+  error message if the plugin specific configuration is invalid, which will be
+  displayed as the charm's status message, or an empty string if it's valid.
+  The charm will attempt to use the plugin_dict information to request a certificate.
+  The library itself is responsible for calling this function with the appropriate argument,
+  so there is no need to modify or call this function from the top level charm.
+
+
+Once the charm is deployed, the user requesting a certificate is expected to:
+- Pass in an email as a config option
+- Optionally pass in an ACME server as config option
+- Create a secret that contains all of the required configuration options for the plugin.
+- Grant this secret to the charm.
+- Pass in the secret id that was generated as a config option.
```

Comment:
I'd rather those actually be commands instead of vague statements

Reply:
done